### ISSUE TITLE
Randomizer.GetString can return strings that have invalid surrogate pairs

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#4513]
System.String is UTF-16, and not all sequences of `char` values are valid UTF-16. The current implementation disregards this:

https://github.com/nunit/nunit/blob/9b5ec78f28a00077ef4928f3a62a21a6853283e6/src/NUnitFramework/framework/Internal/Randomizer.cs#L502-L510

This is only a concern if `System.Char.IsSurrogate` returns true for any character in the `allowedChars` parameter. Possibly, this could be resolved by throwing NotSupportedException for such inputs.

Alternatively, `allowedChars` could be enumerated with https://learn.microsoft.com/en-us/dotnet/api/system.globalization.stringinfo.gettextelementenumerator (available since .NET Framework 1.1) in order to not split apart char sequences that need to stay together in order to be meaningful. This would be a slower code path, so it would be desirable to still check to see if `System.Char.IsSurrogate` returns true for any character in the `allowedChars` parameter.

(For some target frameworks, there may be a vectorized way to compute the equivalent of `allowedChars.Any(char.IsSurrogate)`, I'm not sure.)

**Comment by manfred-brands** [comment]
I think we should disallow surrogates. They are a niche and for any code with random strings not likely to be used.

We can rename the current `GetString` into a `UnsafeGetString` and make it `private` which does not check its `allowedChars`
Then the overloads which pass in `DefaultStringChars` call `UnsafeGetString` directly.
Create a new `GetString` which checks the passed in `allowedChars` and `throws` if it has an surrogates.

In addition make an NUnit test to ensure that nobody put surrogates in `DefaultStringChars`.
That way we don't affect the performance of 99% of the `GetString` calls.

Alternatively make a `StringRandomizer` derived class which needs `allowedChars` in the constructor.
That way the check only has to be done once. It then has a `GetString` method without an `allowedChars` parameter.

**Comment by jnm2** [comment]
I'm amenable to that. If someone puts emoji in their allowed chars, we can always revisit. 😄

**Comment by manfred-brands** [comment]
To prevent a merge conflict, this should wait for #4512

**Comment by OsirisTerje** [comment]
@manfred-brands 

> We can rename the current GetString into a UnsafeGetString

Will this be a breaking change ?

Is this something we could fix now ?

**Comment by manfred-brands** [comment]
> > We can rename the current GetString into a UnsafeGetString
> 
> Will this be a breaking change ?
> 

Yes. Even thought it is in the `internal` namespace, the code is `public`.

So we might be better of to detect `surrogates` and if there are any use the slower loop as @jnm2 suggested.
If we add an private `SafeGetString` called by the overloads passing in `DefaultStringChars` the overhead should be none for code using those.
